src/graph_visualization.cpp

```diff
@@ -0,0 +1,463 @@
+    }
+  }
+}
+
+void heapify(std::vector<TreeNode> &nodes, int n, int i) {
+  int largest = i;
+  int l = 2 * i + 1;
+  int r = 2 * i + 2;
+
+  if (l < n && nodes[l].id > nodes[largest].id) {
+    largest = l;
+  }
+
+  if (r < n && nodes[r].id > nodes[largest].id) {
+    largest = r;
+  }
+
+  if (largest != i) {
+    std::swap(nodes[i], nodes[largest]);
+    heapify(nodes, n, largest);
+  }
+}
+
+void idHeapSort(std::vector<TreeNode> &nodes) {
+  int n = 0;
+  for (int i = 0; i < nodes.size(); i++) {
+    n++;
+  }
+
+  for (int i = n / 2 - 1; i >= 0; i--) {
+    heapify(nodes, n, i);
+  }
+
+  for (int i = n - 1; i >= 0; i--) {
+    std::swap(nodes[0], nodes[i]);
+    heapify(nodes, i, 0);
+  }
+}
+
+TreeNode getTreeNodeById(std::vector<TreeNode> &nodes, TreeNode::nodeId id) {
+  TreeNode node = {};
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].id == id) {
+      node = nodes[i];
+      break;
+    }
+  }
+  return node;
+}
+
+void removeIncidentArcs(std::vector<TreeNode> &nodes, TreeNode::nodeId id) {
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      if (nodes[i].succ[j] == id) {
+        nodes[i].succ.erase(nodes[i].succ.begin() + j);
+      }
+    }
+    for (int j = 0; j < nodes[i].pred.size(); j++) {
+      if (nodes[i].pred[j] == id) {
+        nodes[i].pred.erase(nodes[i].pred.begin() + j);
+      }
+    }
+  }
+}
+
+void greedyFAS(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges) {
+  degReinit(nodes);
+
+  std::vector<TreeNode> copyNodes = nodes;
+  std::vector<TreeNode> s1 = {}, s2 = {};
+
+  while (nodes.size() != 0) {
+    for (int i = 0; i < nodes.size(); i++) {
+      if (nodes[i].degP == 0) {
+        s2.push_back(copyNodes[nodes[i].id]);
+
+        removeIncidentArcs(nodes, nodes[i].id);
+        nodes.erase(nodes.begin() + i);
+        degReinit(nodes);
+
+        i = -1;
+      }
+    }
+
+    for (int i = 0; i < nodes.size(); i++) {
+      if (nodes[i].degM == 0) {
+        s1.push_back(copyNodes[nodes[i].id]);
+
+        removeIncidentArcs(nodes, nodes[i].id);
+        nodes.erase(nodes.begin() + i);
+        degReinit(nodes);
+
+        i = -1;
+      }
+    }
+
+    if (nodes.size() != 0) {
+      int max = nodes[0].degP - nodes[0].degM;
+      int posMax = 0;
+      for (int i = 0; i < nodes.size(); i++) {
+        if (nodes[i].degP - nodes[i].degM > max) {
+          max = nodes[i].degP - nodes[i].degM;
+          posMax = i;
+        }
+      }
+      TreeNode maxNode = getTreeNodeById(copyNodes, nodes[posMax].id);
+      s1.push_back(maxNode);
+
+      removeIncidentArcs(nodes, nodes[posMax].id);
+      nodes.erase(nodes.begin() + posMax);
+      degReinit(nodes);
+    }
+  }
+
+  std::vector<TreeNode> s = {};
+  for (int i = 0; i < s1.size(); i++) {
+    s.push_back(s1[i]);
+  }
+  for (int i = 0; i < s2.size(); i++) {
+    s.push_back(s2[s2.size() - 1 - i]);
+  }
```

Comment:
Again, not a fan of the standard library API
```suggestion
  s.insert(s.end(), s2.rbegin(), s2.rend());
```